### DISCUSSION TITLE
backup and restore - can't just copy .config/autokey?

### DISCUSSION THREAD

**Discussion by joaociocca** [autokey/autokey#952] [Q&A]
~~I tried copying over my `~/.config/autokey` elsewhere, but then when I try to run Autokey I get a "Fatal error starting AutoKey." saying that `~/.config/autokey/data` already exists...?~~

ok, this was caused by a bad symlink... now that's fixed, but when I try to run, new fatal error: "Error while saving configuration. Backup has been restored (if found)."

And yeah, I've seen both #933 and #333 - I used `rsync -a` to copy the original files to my github personal repo, then link back the `autokey` folder to `.config`. On the original computer, it still works after that - but when I clone/sync the repo to a new computer, link `autokey` to its `.config` directory... these errors pop up. =(

**Comment by josephj11** [comment]
Welcome to the AutoKey community, @joaociocca !

Not sure what's going on here, but:

Look in the new $HOME/.config/autokey/autokey.json

Mine has the following two entries:
    "userCodeDir": "/home/shelelia/Autokey/Modules",
...
    "folders": [
        "/home/shelelia/Autokey"

which are hard-coded paths. If your user is different on the new computer, these will break.
Try manually correcting these and see what happens.

Generally, on a new machine, you should install AutoKey normally and run it once so it can initialize itself. Terminate AutoKey normally,  then copy everything from your backup starting at .config/autokey/data, not all the way at the top. Then restart AutoKey.

Try this and see what happens.

Both of those old issues deal with hidden files (which your rsync would have handled correctly). I got the devs to get rid of hidden files in 0.96.0, so if you're on that (our current release), that's a non-issue.

**Comment by joaociocca** [comment]
no `userCodeDir`, and `folders:[]` is empty...

ok, it's running now... let's see about linking my folder into data, then!

nice, this did the job! so no need to backup the whole /autokey folder within .config, just the specific script folder inside /data! Thanks, Joe!

**Reply by josephj11** [reply]
Great!

Since you are already thinking along these lines, I'm including a QA with another user below. I haven't gotten around to adding it to our wiki yet.

Feel free to join us over on [Gitter](https://gitter.im/autokey/autokey) to discuss anything related to AutoKey. It's pretty quiet there lately, but there are a lot of members who may see what you post and respond.

If you haven't been there yet, be sure to check out our [wiki](https://github.com/autokey/autokey/wiki/). There are a number of useful articles and lots of example scripts.

----

I'd like to specify which folder my scripts live in. My use case is simple: I have several different machines I do my work on, and I have to manually maintain the scripts between them. If I could have them in a single common location, that would make my life a lot easier. OR...is that already there and I'm just missing it?

You missed it because it's generic to Linux - and we do have the option.

There are two parts to this. First, within AutoKey, when you add a new folder in the folder pane, you have the option to Create elsewhere. This can be anywhere on your system as long as it will be available/mounted when AutoKey is running.

I have had all my actions (phrases and scripts) in an external subdirectory tree for years. This is great because it makes backups easier and keeps my actions intact if I ever have to reinstall or even purge AutoKey.

If you use this option, you should add a new empty folder from within AutoKey and then move all your actions into it while AutoKey is not running. This avoids a lot of potential problems. Once you do this, AutoKey "owns" this folder/subtree. If you delete it in AutoKey it gets deleted from your file system. If you move it from within AutoKey, it will only move the AutoKey action files. Any other files that happen to be there will be lost.

 The next level of this (which I also use) is that the AutoKey configuration directory, $HOME/.config/autokey can be a symlink. This allows you to move or change the configuration (including all your action files) any time you want to (while AutoKey is not running!). I used this so I could dynamically switch between the current version of AutoKey and the Beta version which had some breaking changes as part of 0.96.0.

This could also be used to switch all your AutoKey actions from one set to another when you change work flows.

If you want to, you can use these methods to put your AutoKey configuration(s)/actions in a shared folder like one that is managed by Dropbox, Mega, or a similar service.

Be aware that AutoKey is not designed to support multiple instances of itself running concurrently, so if you do this, things are likely to break.

If you run these separate instances one at a time, things should work, but this is mostly unexplored territory. One known concern would be that AutoKey updates usage counts in action metadata files so it can support sorting menus by frequency of use. I don't see why this would be a problem, but if you use this sorting option, the results might be unexpected when these counts are modified by one instance of AutoKey and then accessed by another instance.

**Reply by joaociocca** [reply]
> The next level of this (which I also use) is that the AutoKey configuration directory, $HOME/.config/autokey can be a symlink. This allows you to move or change the configuration (including all your action files) any time you want to (while AutoKey is not running!). I used this so I could dynamically switch between the current version of AutoKey and the Beta version which had some breaking changes as part of 0.96.0.

I tried doing exactly this, but this is the one that lead to errors. On the original machine, I was able to move the `~/.config/autokey` folder to my git repo folder, then link it back to `~/.config` and everything just worked. But on the new machine, when I tried linking the git repo folder to `~/.config` led to those errors...

The only way I got it to actually work was, like you said, run Autokey once first to let it create the `~/.config/autokey` folder, and then link just the script folder inside `/data` to it...
